channelserver/src/server.rs

```diff
@@ -0,0 +1,257 @@
+//! `ChannelServer` is an actor. It maintains list of connection client session.
+//! And manages available channels. Peers send messages to other peers in same
+//! channel through `ChannelServer`.
+
+// use std::sync::{Arc, Mutex};
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::time::Instant;
+
+use actix::prelude::{Actor, Context, Handler, Recipient};
+use rand::{self, Rng, ThreadRng};
+use uuid::Uuid;
+
+use logging::MozLogger;
+use perror;
+use settings::Settings;
+
+//pub type ChannelCollection = Arc<Mutex<HashMap<Uuid, HashMap<usize, Channel>>>>;
+//pub type SessionCollection = Arc<Mutex<HashMap<usize,Recipient<TextMessage>>>>;
+
+/// Chat server sends this messages to session
+#[derive(Message)]
+pub struct TextMessage(pub String);
+
+/// Message for chat server communications
+
+/// New chat session is created
+#[derive(Message)]
+#[rtype(usize)]
+pub struct Connect {
+    pub addr: Recipient<TextMessage>,
+    pub channel: Uuid,
+}
+
+/// Session is disconnected
+#[derive(Message)]
+pub struct Disconnect {
+    pub channel: Uuid,
+    pub id: usize,
+}
+
+/// Send message to specific channel
+#[derive(Message)]
+pub struct ClientMessage {
+    /// Id of the client session
+    pub id: usize,
+    /// Peer message
+    pub msg: String,
+    /// channel name
+    pub channel: Uuid,
+}
+
+#[derive(Hash, Eq, PartialEq, Clone, Debug)]
+pub struct Channel {
+    pub id: usize,
+    pub started: Instant,
+    pub msg_count: u8,
+    pub data_exchanged: usize,
+}
+
+/// `ChannelServer` manages chat channels and responsible for coordinating chat
+/// session. implementation is super primitive
+pub struct ChannelServer {
+    me: usize, // id (to see if threading an issue)
+    // collections of sessions grouped by channel
+    channels: HashMap<Uuid, HashMap<usize, Channel>>,
+    // individual connections
+    sessions: HashMap<usize, Recipient<TextMessage>>,
+    rng: RefCell<ThreadRng>,
+    log: MozLogger,
+    pub settings: RefCell<Settings>,
+}
+
+impl Default for ChannelServer {
+    fn default() -> ChannelServer {
+        let me = rand::thread_rng().gen::<usize>();
+        ChannelServer {
+            me: me,
+            channels: HashMap::new(),
+            sessions: HashMap::new(),
+            rng: RefCell::new(rand::thread_rng()),
+            log: MozLogger::default(),
+            settings: RefCell::new(Settings::new().unwrap()),
+        }
+    }
+}
+
+impl ChannelServer {
+    /// Send message to all users in the channel except skip_id
+    fn send_message(
+        &mut self,
+        channel: &Uuid,
+        message: &str,
+        skip_id: usize,
+    ) -> Result<(), perror::HandlerError> {
+        if let Some(participants) = self.channels.get_mut(channel) {
+            // show's over, everyone go home.
+            if message == "\04" {
+                for (id, info) in participants {
```

Comment:
This "arbitrarily many participants in a channel" thing feels like YAGNI from my perspective.  Do we have a concrete use-case for it?

For the purposes of our pairing flow, I think it would be more useful to error out if more than two participants try to join a single channel, as that's an indication that something unexpected is happening.

Reply:
Honestly, this was more an element of my own laziness. The original code allowed a large number of participants so it would be more work to remove the code to allow that than to just leave it in place and restrict the per channel participation.